lib/r10k/source/exec.rb

```diff
@@ -0,0 +1,45 @@
+require 'open3'
+require 'json'
+require 'yaml'
+
+class R10K::Source::Exec < R10K::Source::Hash
+  R10K::Source.register(:exec, self)
+
+  def initialize(name, basedir, options = {})
+    unless command = options[:command]
+      raise ConfigError, _('Environment source %{name} missing required parameter: command') % {name: name}
+    end
+
+    begin
+      stdout, stderr, status = Open3.capture3(command)
+    rescue Errno::EACCES, Errno::ENOENT => e
+      raise R10K::Error, _("Error executing command for exec source %{name}: %{msg}") % {name: name, msg: e.message}
+    end
+
+    unless status.success?
+      raise R10K::Error, _("Error executing exec command for source %{name}: %{cmd}. Command exited %{code}\n" \
+                           "stdout:\n%{out}\n" \
+                           "stderr:\n%{err}") % {name: name, cmd: command, code: status.exitstaus.to_s, out: stdout, err: stderr}
+    end
+
+    begin
```

Comment:
Supporting two different serialization formats adds complexity, can we start with JSON support to start with and add YAML support if the demand manifests for that format?

If we do need to support multiple serialization formats, that should be configured at the source level, with something like this:

```
sources:
  exec:
    command: "..."
    format: "yaml"
```

Reply:
We could start with JSON. That is a choice we could make.

Design-wise, regardless of what we decide we should not let the user specify. We can figure out JSON vs. YAML trivially so there's never going to be a benefit to making the user think about it or care during configuration. We should optimize for keeping the configuration simple.

My design preference is to support both. YAML is prevalent in the R10K ecosystem, and is the input format accepted by the conceptually related ENC feature in Puppet. However, the use case that's driving this feature is to retrieve environment information from REST APIs, which typically return JSON. The code complexity to just accept both is minimal. This is effectively a user interface so anything we can do to make dealing with it simpler for the user is a win.

Reply:
To account for yaml being more permissive than JSON it would be best to implement validation of the data read though. For today/tonight I added the following pending test, but the PR doesn't currently implement it.

```ruby
it "rejects schema invalid data"
```